velocidi-sdk/src/main/kotlin/com/velocidi/Velocidi.kt

```diff
@@ -1,5 +1,91 @@
 package com.velocidi
 
-fun main(args : Array<String>) {
-    println("Welcome to Velocidi SDK")
+import android.Manifest
+import android.content.Context
+import android.util.Log
+import com.android.volley.Request
+import org.json.JSONObject
+import java.util.*
+
+
+class Velocidi constructor(val config: Config, val adInfo: AdvertisingInfo, context: Context) {
+
+    private val client = HttpClient()
+
+    private val appInfo = Util.getApplicationInfo(context)
+
+    init {
+        client.headers["User-Agent"] = "${appInfo.appName}/${appInfo.appVersion} ${appInfo.androidSDK} ${appInfo.device}"
+        client.defaultParams["aaid"] = adInfo.id
+        client.defaultParams["cookies"] = "false"
+    }
+
+     private fun handleRequest(req: com.velocidi.Request) {
+        if(!adInfo.shouldTrack) return
+
+         return when(req) {
+            is com.velocidi.Request.TrackRequest ->
+                if(config.track.enabled)
+                    client.sendRequest(Request.Method.POST, config.track.host.toString(), req.attributes)
+                else return
+            is com.velocidi.Request.MatchRequest ->
+                if(config.match.enabled)
+                    client.sendRequest(Request.Method.GET, "${config.match.host}?${req.toQueryParams()}")
+                else return
+
+        }
+    }
+
+     companion object {
+         internal lateinit var instance: Velocidi
+
+         internal val queue: Queue<com.velocidi.Request> = FixedSizeQueue(300)
+
+         fun start(config: Config, context: Context) {
+             if(!Util.checkPermission(context, Manifest.permission.INTERNET))
+                 Log.e(Constants.LOG_TAG, "Velocidi SDK requires Internet permission")
+
+             if(::instance.isInitialized)
+                 return
+
+             val listener = object : AdvertisingIdListener {
+                 override fun fetchAdvertisingIdCompleted(advertisingInfo: AdvertisingInfo) {
+                     instance = Velocidi(config, advertisingInfo, context)
```

Comment:
I totally agree with you. I tried something similar before but this approach makes testing way harder. Do you have some suggestions for the tests?

Reply:
Does this make testing harder because you want to mock the GetAdvertisingIdTask? 

If so, we can do something like this:

```kotlin
abstract class LoadAdvertisinId<T>(val context: Context) {
    lateinit var advertisingInfo: AdvertisingInfo

    init {
        fetchAndSetAdvertisingId()
    }

	fun fetchAndSetAdvertisingId(){
		val listener = object : AdvertisingIdListener {
            override fun fetchAdvertisingIdCompleted(advertisingInfo: AdvertisingInfo) {
                this@LoadAdvertisinId.advertisingInfo = advertisingInfo

                while(queue.size != 0){
                    handleTask(queue.remove())
                }
            }
        }
        GetAdvertisingIdTask(listener).execute(context)
	}
}
```

We can then create a `MockVelocidi` that extends `Velocidi` and overrides the `fetchAndSetAdvertisingId`

```kotlin

override fun fetchAndSetAdvertisingId(){
	advertisingInfo = "randomAdId"
}
```

Reply:
Check this 5f5625c.

I think this logic is getting a bit messy. We are calling a non-final function in the constructor which is a dangerous operation. I also don't like having two 2 callback for with the same objective (`fetchAndSetAdvertisingId` and `onFetchCompleted`).

I prefer the old approach where we delay the initialization of Velocidi.
I see 3 advantages with this approach:
- The class already has all the information needed to process match and track requests. We don't have two worry about any async tasks inside the class or with the superclass. Only in the `start` function of the companion object.
- Makes testing a bit easier in my opinion
- We don't instantiate nor do any kind of processing(creating the Http client, setting up cache and request queue) if the retrieval of the AAID fails.

Regarding this:
> This will propagate null unsafety all throughout the code and may lead to unexpected NullPointExceptions. 

It is unlikely since `getInstance` is a private function and only track and match function called it with a `safe call operator` operator.

What do you think @duartepinto @belerophon ?

Reply:
Are the client applications also delayed with that approach?

In general, I prefer an asynchronous approach to this. But, in this case in particular, I think we must always go for the approach that leaves a greater amount of free resources for the client application at any given time.

Reply:
Just to write-down what we discussed offline:
- we provide two "static" methods:
  - one `.init()` which the developer can use to configure the SDK; it can return the instance immediatelly;
  - one `.getInstance()` which can always be used to get the current instance; it should throw a proper exception if the SDK has not been initialized before.
- the `track` and `match` methods are only available as part of the Velocidi object instance.